hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapReplicationStateHolder.java

```diff
@@ -174,18 +181,34 @@ void applyState() {
 
                 long nowInMillis = Clock.currentTimeMillis();
 
+                long ownedEntryCountOnThisNode = entryCountOnThisNode(mapContainer);
+                EvictionConfig evictionConfig = mapContainer.getMapConfig().getEvictionConfig();
+                boolean perNodeEvictionConfigured = mapContainer.getEvictor() != Evictor.NULL_EVICTOR
+                        && evictionConfig.getMaxSizePolicy() == PER_NODE;
+
                 for (int i = 0; i < keyRecord.size(); i += 2) {
                     Data dataKey = (Data) keyRecord.get(i);
                     Record record = (Record) keyRecord.get(i + 1);
 
-                    recordStore.putReplicatedRecord(dataKey, record, nowInMillis, populateIndexes);
-
-                    if (recordStore.shouldEvict()) {
-                        // No need to continue replicating records anymore.
-                        // We are already over eviction threshold, each put record will cause another eviction.
-                        recordStore.evictEntries(dataKey);
-                        break;
+                    if (perNodeEvictionConfigured) {
+                        if (ownedEntryCountOnThisNode >= evictionConfig.getSize()) {
+                            if (operation.getReplicaIndex() == 0) {
+                                recordStore.doPostEvictionOperations(dataKey, record);
+                            }
+                        } else {
+                            recordStore.putReplicatedRecord(dataKey, record, nowInMillis, populateIndexes);
+                            ownedEntryCountOnThisNode++;
+                        }
+                    } else {
+                        recordStore.putReplicatedRecord(dataKey, record, nowInMillis, populateIndexes);
+                        if (recordStore.shouldEvict()) {
```

Comment:
Curious, how is this check different from the one you introduced? And how come we do post-eviction operations in the first case, but not here?

Reply:
>Curious, how is this check different from the one you introduced?

`shouldEvict` check partition size, in my solution, i scan all partitions to find total number of entries on node.

>And how come we do post-eviction operations in the first case, but not here?

`recordStore.evictEntries` does post-eviction ops.

Reply:
Ok, so it's about different eviction strategies. Thanks